versioned/mongodb/src/main/java/com/dremio/nessie/versioned/store/mongodb/codecs/BsonToEntityConverter.java

```diff
@@ -0,0 +1,158 @@
+/*
+ * Copyright (C) 2020 Dremio
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.dremio.nessie.versioned.store.mongodb.codecs;
+
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+
+import org.bson.BsonReader;
+import org.bson.BsonType;
+
+import com.dremio.nessie.versioned.store.Entity;
+import com.dremio.nessie.versioned.store.mongodb.MongoDbConstants;
+
+/**
+ * Converter for serializing a BSON object to an Entity.
+ */
+public class BsonToEntityConverter {
+  private static final String MONGO_ID_NAME = "_id";
+
+  /**
+   * Deserializes a Entity attributes from the BSON stream.
+   *
+   * @param reader the reader to deserialize the Entity attributes from.
+   * @return the Entity attributes.
+   */
+  public static Map<String, Entity> read(BsonReader reader) {
+    final BsonType type = reader.getCurrentBsonType();
+    if (BsonType.DOCUMENT != type) {
+      throw new UnsupportedOperationException(
+          String.format("BSON serialized data must be a document at the root, type is %s",type));
+    }
+
+    return readDocument(reader);
+  }
+
+  /**
+   * Read a BSON document and deserialize to associated Entity types.
+   * @param reader the reader to deserialize the Entities from.
+   * @return The deserialized collection of entities with their names.
+   */
+  private static Map<String, Entity> readDocument(BsonReader reader) {
+    reader.readStartDocument();
+    final Map<String, Entity> attributes = new HashMap<>();
+
+    while (BsonType.END_OF_DOCUMENT != reader.readBsonType()) {
+      final String name = reader.readName();
+      if (name.equals(MONGO_ID_NAME)) {
+        reader.skipValue();
+        continue;
+      }
+
+      switch (reader.getCurrentBsonType()) {
+        case DOCUMENT:
+          attributes.put(name, Entity.ofMap(readDocument(reader)));
+          break;
+        case ARRAY:
+          attributes.put(name, readArray(reader));
+          break;
+        case BOOLEAN:
+          attributes.put(name, Entity.ofBoolean(reader.readBoolean()));
+          break;
+        case STRING: {
+          final String value = reader.readString();
+          if (value.startsWith(MongoDbConstants.STRING_PREFIX)) {
```

Comment:
Let's remove the string prefix behavior. It is going to make things like future range conditions on numbers a mess. If range is a problem, let's come up with what our number Entity range is instead of this hacking.